frontend/src/Connection.ts

```diff
@@ -565,3 +588,26 @@ export class Connection {
     this.bindSocket();
   };
 }
+
+/**
+ * Fire a function if the timer runs out. You can reset it, or
+ * cancel it to prevent the function from being fired.
+ *
+ * @param onExpired
+ * @param timeoutMs
+ * @returns
+ */
+function resettableTimeout(onExpired: () => void, timeoutMs: number) {
```

Comment:
equivalent to `setTimeout`, but it can be reset as well as cancelled

Reply:
```suggestion
function resettableTimeout(onExpired: () => void, timeoutMs: Types.Milliseconds) {
```

Not sure how annoying that is with the quadratic expansion function, `number` is fine if it is.